resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosHadoopDelegationTokenManager.scala

```diff
@@ -0,0 +1,185 @@
+        logError("Failed to initialize Hadoop delegation tokens\n" +
+          s"\tPricipal: $principal\n\tmode: $mode\n\tsecret file $secretFile\n\tException: $e")
+        throw e
+    }
+  }
+
+  scheduleTokenRenewal()
+
+  private def getSecretFile(conf: SparkConf): (Option[String], String) = {
+    val keytab = conf.get(config.KEYTAB)
+    val tgt = Option(conf.getenv(SparkHadoopUtil.TICKET_CACHE_ENVVAR))
+    val (secretFile, mode) = if (keytab.isDefined && tgt.isDefined) {
+      // if a keytab and a specific ticket cache is specified use the keytab and log the behavior
+      logWarning(s"Keytab and TGT were detected, using keytab, ${keytab.get}, " +
+        s"unset ${config.KEYTAB.key} to use TGT (${tgt.get})")
+      (keytab, "keytab")
+    } else {
+      val m = if (keytab.isDefined) "keytab" else "tgt"
+      val sf = if (keytab.isDefined) keytab else tgt
+      (sf, m)
+    }
+
+    if (principal == null) {
+      require(mode == "tgt", s"Must specify a principal when using a Keytab, was $principal")
+      logInfo(s"Using ticket cache to fetch Hadoop delegation tokens")
+    } else {
+      logInfo(s"Using principal: $principal with mode and keytab $keytab " +
+        s"to fetch Hadoop delegation tokens")
+    }
+
+    logDebug(s"secretFile is $secretFile")
+    (secretFile, mode)
+  }
+
+  private def scheduleTokenRenewal(): Unit = {
+    def scheduleRenewal(runnable: Runnable): Unit = {
+      val remainingTime = timeOfNextRenewal - System.currentTimeMillis()
+      if (remainingTime <= 0) {
+        logInfo("Credentials have expired, creating new ones now.")
+        runnable.run()
+      } else {
+        logInfo(s"Scheduling login from keytab in $remainingTime millis.")
+        credentialRenewerThread.schedule(runnable, remainingTime, TimeUnit.MILLISECONDS)
+      }
+    }
+
+    val credentialRenewerRunnable =
+      new Runnable {
+        override def run(): Unit = {
+          try {
+            val tokensBytes = getNewDelegationTokens
+            broadcastDelegationTokens(tokensBytes)
+          } catch {
+            case e: Exception =>
+              // Log the error and try to write new tokens back in an hour
+              logWarning("Couldn't broadcast tokens, trying again in an hour", e)
+              credentialRenewerThread.schedule(this, 1, TimeUnit.HOURS)
+              return
+          }
+          scheduleRenewal(this)
+        }
+      }
+    scheduleRenewal(credentialRenewerRunnable)
+  }
+
+  private def getNewDelegationTokens(): Array[Byte] = {
+    logInfo(s"Attempting to login to KDC with ${conf.get(config.PRINCIPAL).orNull}")
+    // Get new delegation tokens by logging in with a new UGI
+    // inspired by AMCredentialRenewer.scala:L174
+    val ugi = if (mode == "keytab") {
+      UserGroupInformation.loginUserFromKeytabAndReturnUGI(principal, secretFile.get)
+    } else {
+      // if the ticket cache is not explicitly defined, use the default
+      if (secretFile.isEmpty) {
```

Comment:
The point I was trying to make is that you do not need *any* special handling for TGT. The UGI class already does everything you need, you just need to get the current user. It will keep the TGT updated with any changes that happen on disk. You don't need to handle `KRB5CCNAME` anywhere, because UGI should be doing that for you. If it's not, you need to explain why you need this special handling, because the expected behavior is for this to work without you needing to do anything.

So you can simplify this class by *only* handling the principal / keytab case, and just using `UserGroupInformation.getCurrentUser` in the other case. You don't need to keep track of the "mode" or anything else, just whether you're using a principal / keytab pair.

Reply:
How do the executors get the TGT though?  `KRB5CCNAME` would need to be set in the executor containers as well, right? If it is, I suppose you don't need to broadcast the delegation tokens at all because `runAsSparkUser` takes care of it for you? 

I guess now that https://github.com/apache/spark/pull/19437 has been merged we can disseminate the TGT though Mesos.

Reply:
> How do the executors get the TGT though? 

They don't. That's why you're creating delegation tokens and sending *them* to the executor.

Reply:
Ah gotcha. I misunderstood your previous comment to mean that you wouldn’t need to renew the tokens when using the ticket cache. I’ll simplify the logic. Thanks.

Reply:
So `getCurrentUser` actually doesn't work. I believe for the reason mentioned [here](https://github.com/apache/spark/blob/a18d637112b97d2caaca0a8324bdd99086664b24/resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/security/AMCredentialRenewer.scala#L152). The effect is that the same expired credentials are returned (causing the renewer to loop). 

I understand that using the TGT instead of a keytab is different than the YARN reference implementation, and it's unusual to use it in this case since it expires anyways. Do you think it would be better to avoid the ticket renewal logic all together when using a TGT or to keep the older `UserGroupInformation.getUGIFromTicketCache`-based method?

Reply:
> Do you think it would be better to avoid the ticket renewal logic all together when using a TGT

If `getCurrentUser` does not work, then yes, that's probably the best way forward. Requiring the user to set `KRB5CCNAME` is not really a good way to go about this. Also because in that case the user still has to make sure the TGT is updated through other means.

Reply:
Agreed, thanks.